diff -ruNP ucarp-1.2.orig/configure ucarp-1.2/configure
--- ucarp-1.2.orig/configure	2006-02-25 17:32:42.000000000 +0100
+++ ucarp-1.2/configure	2007-03-10 15:10:20.000000000 +0100
@@ -9231,7 +9231,6 @@
   echo $ECHO_N "(cached) $ECHO_C" >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lpcap  $LIBS"
 cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
 _ACEOF
@@ -9294,7 +9293,6 @@
 #define HAVE_LIBPCAP 1
 _ACEOF
 
-  LIBS="-lpcap $LIBS"
 
 fi
 
--- ucarp-1.2.orig/src/carp.c	2006-02-19 00:06:17.000000000 +0100
+++ ucarp-1.2/src/carp.c	2007-03-14 19:52:17.000000000 +0100
@@ -48,6 +48,22 @@
 # include <dmalloc.h>
 #endif
 
+static inline void tv_fix_usec_overrun(struct timeval *tv)
+{
+	while (tv->tv_usec >= 1000000) {
+		tv->tv_usec-=1000000;
+		tv->tv_sec++;
+	}
+}
+
+static inline void tv_fix_usec_underrun(struct timeval *tv)
+{
+	while (tv->tv_usec < 0) {
+		tv->tv_usec+=1000000;
+		tv->tv_sec--;
+	}
+}
+
 static void carp_set_state(struct carp_softc *sc, int state)
 {
     if ((int) sc->sc_state == state) {
@@ -172,6 +188,16 @@
     return 0;
 }
 
+static inline void make_mcast_ether_from_ipv4(struct in_addr addr, unsigned char *ether)
+{
+	ether[0] = 0x01;
+	ether[1] = 0x00;
+	ether[2] = 0x5e;
+	ether[3] = (ntohl(addr.s_addr) & 0x007f0000) >> 16;
+	ether[4] = (ntohl(addr.s_addr) & 0x0000ff00) >> 8;
+	ether[5] = (ntohl(addr.s_addr) & 0x000000ff);
+}
+
 static void carp_send_ad(struct carp_softc *sc)
 {
     struct carp_header ch;
@@ -214,6 +240,7 @@
         logfile(LOG_ERR, _("Out of memory to create packet"));
         sc->sc_ad_tmo.tv_sec = now.tv_sec + tv.tv_sec;
         sc->sc_ad_tmo.tv_usec = now.tv_usec + tv.tv_usec;            
+	tv_fix_usec_overrun(&sc->sc_ad_tmo);
         return;
     }
     ip.ip_v = IPVERSION;
@@ -243,12 +270,7 @@
     eh.ether_shost[4] = 0x00;
     eh.ether_shost[5] = vhid;
     
-    eh.ether_dhost[0] = 0xff;
-    eh.ether_dhost[1] = 0xff;
-    eh.ether_dhost[2] = 0xff;
-    eh.ether_dhost[3] = 0xff;
-    eh.ether_dhost[4] = 0xff;
-    eh.ether_dhost[5] = 0xff;        
+    make_mcast_ether_from_ipv4(ip.ip_dst, eh.ether_dhost);
     
     eh.ether_type = htons(ETHERTYPE_IP);    
     
@@ -298,6 +320,7 @@
     if (advbase != 255 || advskew != 255) {
         sc->sc_ad_tmo.tv_sec = now.tv_sec + tv.tv_sec;
         sc->sc_ad_tmo.tv_usec = now.tv_usec + tv.tv_usec;            
+	tv_fix_usec_overrun(&sc->sc_ad_tmo);
         /* IPv6 ? */        
     }
 }
@@ -332,19 +355,23 @@
         case AF_INET:
             sc->sc_md_tmo.tv_sec = now.tv_sec + tv.tv_sec;
             sc->sc_md_tmo.tv_usec = now.tv_usec + tv.tv_usec;            
+	    tv_fix_usec_overrun(&sc->sc_md_tmo);
             break;
 #ifdef INET6
         case AF_INET6:
             sc->sc_md6_tmo.tv_sec = now.tv_sec + tv.tv_sec;
             sc->sc_md6_tmo.tv_usec = now.tv_usec + tv.tv_usec;            
+	    tv_fix_usec_overrun(&sc->sc_md6_tmo);
             break;
 #endif /* INET6 */
         default:
             sc->sc_md_tmo.tv_sec = now.tv_sec + tv.tv_sec;
             sc->sc_md_tmo.tv_usec = now.tv_usec + tv.tv_usec;
+	    tv_fix_usec_overrun(&sc->sc_md_tmo);
 #ifdef INET6
             sc->sc_md6_tmo.tv_sec = now.tv_sec + tv.tv_sec;
             sc->sc_md6_tmo.tv_usec = now.tv_usec + tv.tv_usec;
+	    tv_fix_usec_overrun(&sc->sc_md6_tmo);
 #endif
             break;
         }        
@@ -354,6 +381,7 @@
         tv.tv_usec = (unsigned int) (sc->sc_advskew * 1000000ULL / 256ULL);
         sc->sc_md_tmo.tv_sec = now.tv_sec + tv.tv_sec;
         sc->sc_md_tmo.tv_usec = now.tv_usec + tv.tv_usec;
+	tv_fix_usec_overrun(&sc->sc_md_tmo);
         /* No IPv6 scheduling ? */
         break;
     }
@@ -384,43 +412,23 @@
     }    
 }
 
-static void packethandler(unsigned char *dummy,
-                          const struct pcap_pkthdr *header,
-                          const unsigned char *sp)
+static void packethandler(int fd)
 {
-    struct ether_header etherhead;
     struct ip iphead;
     unsigned int source;
     unsigned int dest;
     unsigned char proto;
     unsigned int caplen;
     unsigned int ip_len;
-            
-    (void) dummy;
-    if (header->caplen <= (sizeof etherhead + sizeof iphead)) {
-        return;
-    }    
-    memcpy(&etherhead, sp, sizeof etherhead);
-#ifdef DEBUG
-    logfile(LOG_DEBUG, "Ethernet "
-             "[%02x:%02x:%02x:%02x:%02x:%02x]->[%02x:%02x:%02x:%02x:%02x:%02x] "
-             "type [%04x]",
-            (unsigned int) etherhead.ether_shost[0],
-            (unsigned int) etherhead.ether_shost[1],
-            (unsigned int) etherhead.ether_shost[2],
-            (unsigned int) etherhead.ether_shost[3],
-            (unsigned int) etherhead.ether_shost[4],
-            (unsigned int) etherhead.ether_shost[5],
-            (unsigned int) etherhead.ether_dhost[0],
-            (unsigned int) etherhead.ether_dhost[1],
-            (unsigned int) etherhead.ether_dhost[2],
-            (unsigned int) etherhead.ether_dhost[3],
-            (unsigned int) etherhead.ether_dhost[4],
-            (unsigned int) etherhead.ether_dhost[5],
-            (unsigned int) ntohs(etherhead.ether_type));
-#endif
-    sp += sizeof etherhead;
-    caplen = header->caplen - sizeof etherhead;
+    char buf[1500];
+    unsigned char *sp = buf;
+
+    if ((int)(caplen = recv(fd, buf, sizeof(buf), 0)) < 0)
+	    return;
+
+    if (caplen < sizeof(iphead))
+	    return;
+
     memcpy(&iphead, sp, sizeof iphead);    
     if (iphead.ip_src.s_addr == srcip.s_addr) {
 	return;
@@ -635,10 +643,12 @@
 
 int docarp(void)
 {
-    struct bpf_program bpfp;
     struct pollfd pfds[1];
+    fd_set fds, rfds;
     int nfds;
-    char errbuf[PCAP_ERRBUF_SIZE];
+    struct ifreq ifr;
+    int mcast_fd;
+    struct ip_mreqn mreq;
 
     sc.sc_vhid = vhid;
     sc.sc_advbase = advbase;
@@ -674,23 +684,53 @@
             (unsigned int) hwaddr[0], (unsigned int) hwaddr[1],
             (unsigned int) hwaddr[2], (unsigned int) hwaddr[3],
             (unsigned int) hwaddr[4], (unsigned int) hwaddr[5]);
-    if ((dev_desc = pcap_open_live(interface, ETHERNET_MTU, 0,
-                                   CAPTURE_TIMEOUT, errbuf)) == NULL) {
-        logfile(LOG_ERR, _("Unable to open interface [%s]: %s"),
-                interface, errbuf);
-        return -1;
-    }    
-    if (pcap_compile(dev_desc, &bpfp, build_bpf_rule(),
-                     1, (bpf_u_int32) 0) != 0) {
-        logfile(LOG_ERR, _("Unable to compile pcap rule: %s [%s]"),
-                errbuf, interface);
-        return -1;
+    if ((dev_desc_fd = socket(PF_PACKET, SOCK_RAW, 0)) == -1) {
+	    logfile(LOG_ERR, _("socket() error"));
+	    return -1;
+    }
+    {
+	    struct sockaddr_ll sa;
+	    strncpy(ifr.ifr_name, interface, sizeof(ifr.ifr_name));
+
+	    if (ioctl(dev_desc_fd, SIOCGIFINDEX, &ifr) == -1) {
+		    logfile(LOG_ERR, _("ioctl() error"));
+		    return -1;
+	    }
+
+	    memset(&sa, 0, sizeof(sa));
+
+	    sa.sll_family = AF_PACKET;
+	    sa.sll_ifindex = ifr.ifr_ifindex;
+
+	    if (bind(dev_desc_fd, (struct sockaddr*)&sa, sizeof(sa)) == -1) {
+		    logfile(LOG_ERR, _("bind() error"));
+		    return -1;
+	    }
+    }
+
+    mcast_fd = socket(PF_INET, SOCK_RAW, IPPROTO_VRRP);
+    if (fcntl(mcast_fd, F_SETFD, FD_CLOEXEC) == -1) {
+	    logfile(LOG_ERR, _("fcntl(FD_CLOEXEC): %s"), strerror(errno));
+	    return 1;
+    }
+    mreq.imr_multiaddr.s_addr = INADDR_CARP_GROUP;
+    mreq.imr_address.s_addr = srcip.s_addr;
+    mreq.imr_ifindex = ifr.ifr_ifindex;
+    if (setsockopt(mcast_fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq)) == -1) {
+	    logfile(LOG_ERR, _("setsockopt(IP_ADD_MEMBERSHIP): %s"), strerror(errno));
+	    return -1;
+    }
+    if (setsockopt(mcast_fd, SOL_SOCKET, SO_BINDTODEVICE, interface, strlen(interface)+1) == -1) {
+	    logfile(LOG_ERR, _("setsockopt(SO_BINDTODEVICE): %s"), strerror(errno));
+	    return -1;
     }
-    pcap_setfilter(dev_desc, &bpfp);
-    dev_desc_fd = pcap_fileno(dev_desc);
-    pfds[0].fd = dev_desc_fd;
-    pfds[0].events = POLLIN | POLLERR | POLLHUP | POLLNVAL;
     
+    FD_ZERO(&rfds);
+    FD_SET(mcast_fd, &rfds);
+
+    pfds[0].fd = mcast_fd;
+    pfds[0].events = POLLIN | POLLERR | POLLHUP | POLLNVAL;
+
     if (shutdown_at_exit != 0) {
         (void) signal(SIGINT, sighandler_exit);
         (void) signal(SIGQUIT, sighandler_exit);
@@ -704,36 +744,44 @@
     carp_setrun(&sc, 0); 
     
     for (;;) {
-        nfds = poll(pfds, (nfds_t) 1, sc.sc_advbase * 1000);
-        if (nfds == -1 || 
-            (pfds[0].revents & (POLLERR | POLLHUP | POLLNVAL)) != 0) {
-#ifdef DEBUG
-            logfile(LOG_NOTICE, "poll() error");
-#endif
-            break;
-        }
-        if (gettimeofday(&now, NULL) != 0) {
-            logfile(LOG_WARNING, "gettimeofday() failed: %s",
-                    strerror(errno));
-            continue;
-        }        
-        if (nfds == 1) {
-            pcap_dispatch(dev_desc, 1, packethandler, NULL);
-        }
-        if (sc.sc_md_tmo.tv_sec != 0 && timercmp(&now, &sc.sc_md_tmo, >)) {
-            carp_master_down(&sc);
-        }
-#ifdef INET6
-        if (sc.sc_md6_tmo.tv_sec != 0 && timercmp(&now, &sc.sc_md6_tmo, >)) {
-            carp_master_down(&sc);
-        }
-#endif
-        if (sc.sc_ad_tmo.tv_sec != 0 && timercmp(&now, &sc.sc_ad_tmo, >)) {
-            carp_send_ad(&sc);
-        }
+	    struct timeval sel_tv;
+	    fds = rfds;
+
+	    switch (sc.sc_state) {
+		    case MASTER:
+			    if (timercmp(&sc.sc_ad_tmo, &now, <=)) {
+				    carp_send_ad(&sc);
+				    continue; // restart loop
+			    }
+			    sel_tv = sc.sc_ad_tmo;
+			    break;
+		    case BACKUP:
+			    if (timercmp(&sc.sc_md_tmo, &now, <=)) {
+				    carp_master_down(&sc);
+				    continue; // restart loop
+			    }
+			    sel_tv = sc.sc_md_tmo;
+			    break;
+		    default:
+			    logfile(LOG_ERR, "sc_state is invalid");
+			    return 1;
+	    }
+
+	    sel_tv.tv_sec -= now.tv_sec;
+	    sel_tv.tv_usec -= now.tv_usec;
+	    tv_fix_usec_underrun(&sel_tv);
+
+	    nfds = select(mcast_fd+1, &fds, NULL, NULL, &sel_tv);
+	    gettimeofday(&now, NULL);
+
+	    if (nfds < 0) {
+		    logfile(LOG_ERR, "select() failed: %s", strerror(errno));
+		    return 1;
+	    }
+	    if (nfds == 1) {
+		    packethandler(mcast_fd);
+	    }
     }
-    pcap_close(dev_desc);
-    pcap_freecode(&bpfp);
     
     return 0;
 }
diff -ruNP ucarp-1.2.orig/src/carp_p.h ucarp-1.2/src/carp_p.h
--- ucarp-1.2.orig/src/carp_p.h	2004-06-20 11:24:05.000000000 +0200
+++ ucarp-1.2/src/carp_p.h	2007-03-10 15:12:32.000000000 +0100
@@ -3,7 +3,7 @@
 
 static struct carp_softc sc;
 static struct timeval now;
-static pcap_t *dev_desc;
+//static pcap_t *dev_desc;
 static int dev_desc_fd = -1;
 static int carp_suppress_preempt;
 
diff -ruNP ucarp-1.2.orig/src/ucarp.c ucarp-1.2/src/ucarp.c
--- ucarp-1.2.orig/src/ucarp.c	2006-02-21 00:06:08.000000000 +0100
+++ ucarp-1.2/src/ucarp.c	2007-03-10 15:11:59.000000000 +0100
@@ -202,12 +202,8 @@
     }
 #endif    
     if (interface == NULL || *interface == 0) {        
-        interface = pcap_lookupdev(NULL);
-        if (interface == NULL || *interface == 0) {
-            logfile(LOG_ERR, _("You must supply a network interface"));
-            return 1;
-        }
-        logfile(LOG_INFO, _("Using [%s] as a network interface"), interface);
+        logfile(LOG_ERR, _("You must supply a network interface"));
+        return 1;
     }
     if (vhid == 0) {
         logfile(LOG_ERR, _("You must supply a valid virtual host id"));
--- ucarp-1.2.orig/src/ucarp.h	2004-08-28 18:10:28.000000000 +0200
+++ ucarp-1.2/src/ucarp.h	2007-03-10 15:35:12.000000000 +0100
@@ -60,7 +60,12 @@
 #include <netinet/if_ether.h>
 #include <netinet/ip.h>
 #include <poll.h>
-#include <pcap.h>
+//#include <pcap.h>
+#include <netpacket/packet.h>
+
+#ifndef IPPROTO_VRRP
+#define IPPROTO_VRRP 112
+#endif
 
 #ifdef __sun__
 # define u_int8_t uint8_t
