--- perl-5.8.7.orig/hints/linux.sh	2005-04-04 22:08:31.000000000 +0200
+++ perl-5.8.7/hints/linux.sh	2005-07-26 22:09:40.000000000 +0200
@@ -52,10 +52,10 @@
 # We don't use __GLIBC__ and  __GLIBC_MINOR__ because they
 # are insufficiently precise to distinguish things like
 # libc-2.0.6 and libc-2.0.7.
-if test -L /lib/libc.so.6; then
-    libc=`ls -l /lib/libc.so.6 | awk '{print $NF}'`
-    libc=/lib/$libc
-fi
+#if test -L /lib/libc.so.6; then
+#    libc=`ls -l /lib/libc.so.6 | awk '{print $NF}'`
+#    libc=/lib/$libc
+#fi
 
 # Configure may fail to find lstat() since it's a static/inline
 # function in <sys/stat.h>.
@@ -315,3 +315,5 @@
 	;;
 esac
 EOCBU
+
+#ccflags="$ccflags -DPERL_USE_SAFE_PUTENV"
